xds/internal/balancer/gracefulswitch/gracefulswitch_test.go

```diff
@@ -0,0 +1,1354 @@
+	case <-tcc.NewSubConnCh:
+	}
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an UpdateAddresses() call on the ClientConn")
+	case <-tcc.UpdateAddressesAddrsCh:
+	}
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an RemoveSubConn() call on the ClientConn")
+	case <-tcc.RemoveSubConnCh:
+	}
+	oldCurrent := gsb.balancerCurrent.Balancer.(*buildCallbackBal)
+
+	// ClientConn. Since the callback reports a state READY, this new inline
+	// balancer should be swapped to the current.
+	builder = balancer.Get(buildCallbackBalName)
+	gsb.SwitchTo(builder)
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an UpdateState() call on the ClientConn")
+	case <-tcc.NewStateCh:
+	}
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an NewSubConn() call on the ClientConn")
+	case <-tcc.NewSubConnCh:
+	}
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an UpdateAddresses() call on the ClientConn")
+	case <-tcc.UpdateAddressesAddrsCh:
+	}
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an RemoveSubConn() call on the ClientConn")
+	case <-tcc.RemoveSubConnCh:
+	}
+
+	deadline := time.Now().Add(defaultTestTimeout)
+	// Poll to see if pending was deleted (happens in forked goroutine).
+	for {
+		gsb.mu.Lock()
+		if gsb.balancerPending == nil {
+			gsb.mu.Unlock()
+			break
+		}
+		gsb.mu.Unlock()
+		if time.Now().After(deadline) {
+			t.Fatalf("balancerPending was not deleted as the pending LB reported a READY state, which should switch pending to current")
+		}
+		time.Sleep(time.Millisecond)
+	}
+
+	// The old balancer should be deprecated and any calls from it should be a no-op.
+	oldCurrent.newSubConn([]resolver.Address{}, balancer.NewSubConnOptions{})
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestShortTimeout)
+	defer cancel()
+	select {
+	case <-tcc.NewSubConnCh:
+		t.Fatal("Deprecated LB calling NewSubConn() should not forward up to the ClientConn")
+	case <-ctx.Done():
+	}
+}
+
+const balancerName1 = "mock_balancer_1"
+const balancerName2 = "mock_balancer_2"
+const verifyBalName = "verifyNoSubConnUpdateAfterCloseBalancer"
+const buildCallbackBalName = "callbackInBuildBalancer"
+
+func init() {
+	balancer.Register(bb1{})
+	balancer.Register(bb2{})
+	balancer.Register(vbb{})
+	balancer.Register(bcb{})
+}
```

Comment:
Would the stub balancer make these simpler?
https://github.com/grpc/grpc-go/blob/master/internal/balancer/stub/stub.go

Reply:
Ping?

Reply:
Since I switched getting the builder to accessing through symbol, I think what I have now is good. Let me know if you still would prefer using a stub balancer.